### PR TITLE
[flang][debug] Better handle array lower bound of assumed shape arrays.

### LINKED ISSUE #108633: [flang][debug] Lower bound for assumed shape arrays is ignored.
If an assumed shape array happens to have a lower bound, it is ignored by the generated debug info. See the example below.

```
subroutine show (arr)
  integer, dimension (-2:) :: arr
  arr(2) = 5
  print *, arr
 end subroutine show

program test
  interface
    subroutine show (arr)
      integer, dimension (:) :: arr
    end subroutine show
  end interface
  integer, dimension (15) :: arr
  arr = 7
  call show (arr)
  print *, arr
end program test
```

When stopped at line 4, this is what GDB shows 7 as the value of the arr(2) while it should have been 5.

```
(gdb) p arr(2)
$1 = 7
```

### PR DESCRIPTION
**Pull request by abidh**
As mentioned in #108633, we don't respect the lower bound of the assumed shape arrays if those were specified. It happens in both cases:
1. When caller has non-default lower bound and callee has default
2. When callee has non-default lower bound and caller has default

This PR tries to fix this issue by improving our generation of lower bound attribute on DICompositeTypeAttr. If we see a lower bound in the declaration, we respect that. Note that same function is also used for allocatable/pointer variables. We make sure that we get the lower bound from descriptor in those cases. Please note that DWARF assumes a lower bound of 1 so in many cases we don't need to generate the lower bound.

Fixes #108633.

### CODE REVIEW

File: flang/test/Integration/debug-assumed-shape-array.f90 [hunk 1]
```diff
@@ -1,13 +1,20 @@
...
+! CHECK-DAG: ![[ELEM1]] = !DISubrange(count: !DIExpression(DW_OP_push_object_address, DW_OP_plus_uconst, 32, DW_OP_deref), stride: !DIExpression(DW_OP_push_object_address, DW_OP_plus_uconst, 40, DW_OP_deref))
+! CHECK-DAG: ![[ELEM2]] = !DISubrange(count: !DIExpression(DW_OP_push_object_address, DW_OP_plus_uconst, 56, DW_OP_deref), stride: !DIExpression(DW_OP_push_object_address, DW_OP_plus_uconst, 64, DW_OP_deref))
 
+! CHECK-DAG: !DILocalVariable(name: "arr1"{{.*}}type: ![[TY2:[0-9]+]]{{.*}})
+! CHECK-DAG: ![[TY2]] = !DICompositeType(tag: DW_TAG_array_type{{.*}}elements: ![[ELEMS1:[0-9]+]], dataLocation: !DIExpression(DW_OP_push_object_address, DW_OP_deref))
+! CHECK-DAG: ![[ELEMS1]] = !{![[ELEM11:[0-9]+]], ![[ELEM12:[0-9]+]]}
+! CHECK-DAG: ![[ELEM11]] = !DISubrange(count: !DIExpression(DW_OP_push_object_address, DW_OP_plus_uconst, 32, DW_OP_deref), lowerBound: 3, stride: !DIExpression(DW_OP_push_object_address, DW_OP_plus_uconst, 40, DW_OP_deref))
+! CHECK-DAG: ![[ELEM12]] = !DISubrange(count: !DIExpression(DW_OP_push_object_address, DW_OP_plus_uconst, 56, DW_OP_deref), lowerBound: 4, stride: !DIExpression(DW_OP_push_object_address, DW_OP_plus_uconst, 64, DW_OP_deref))
```
Comment (pawosm-arm) [inline flang/test/Integration/debug-assumed-shape-array.f90:20]: Empty line missing, git reports premature EOF here.

**Review by pawosm-arm — APPROVED** [review]
It does indeed fixes the problem we have encountered.

**Review by jeanPerier — APPROVED** [review]
Thanks

File: flang/test/Integration/debug-assumed-shape-array.f90 (hunk 1 above)
Comment (abidh) [reply flang/test/Integration/debug-assumed-shape-array.f90]: Thanks for catching this. Fixed.

File: flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp [hunk 2]
```diff
@@ -114,10 +114,19 @@ mlir::LLVM::DITypeAttr DebugTypeGenerator::convertBoxedSequenceType(
...
   const unsigned indexSize = dimsSize / 3;
   for ([[maybe_unused]] auto _ : seqTy.getShape()) {
     // For each dimension, find the offset of count, lower bound and stride in
     // the descriptor and generate the dwarf expression to extract it.
+    mlir::Attribute lowerAttr = nullptr;
+    // If declaration has a lower bound, use it.
+    if (declOp && declOp.getShift().size() > index) {
+      if (std::optional<std::int64_t> optint =
+              getIntIfConstant(declOp.getShift()[index]))
+        lowerAttr = mlir::IntegerAttr::get(intTy, llvm::APInt(64, *optint));
```
[the remark this answers is not in the document]
Comment (abidh) [reply flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:130]: Thanks for the review. Indeed the variable lower bound case is not handled and I have added a TODO.

**Pull request merged**